paimon-core/src/main/java/org/apache/paimon/table/system/TagsTable.java

```diff
@@ -217,18 +217,49 @@ public RecordReader<InternalRow> createReader(Split split) {
                 throw new IllegalArgumentException("Unsupported split: " + split.getClass());
             }
             Path location = ((TagsSplit) split).location;
-            LeafPredicate predicate = ((TagsSplit) split).tagName;
+            Predicate predicate = ((TagsSplit) split).tagPredicate;
             TagManager tagManager = new TagManager(fileIO, location, branch);
+            String leafName = "tag_name";
 
             Map<String, Tag> nameToSnapshot = new TreeMap<>();
+            Map<String, Tag> predicateMap = new TreeMap<>();
+            if (predicate != null) {
+                if (predicate instanceof LeafPredicate
+                        && ((LeafPredicate) predicate).function() instanceof Equal
+                        && ((LeafPredicate) predicate).literals().get(0) instanceof BinaryString
+                        && predicate.visit(LeafPredicateExtractor.INSTANCE).get(leafName) != null) {
+                    String equalValue = ((LeafPredicate) predicate).literals().get(0).toString();
+                    if (tagManager.tagExists(equalValue)) {
+                        predicateMap.put(equalValue, tagManager.tag(equalValue));
+                    }
+                }
 
-            if (predicate != null
-                    && predicate.function() instanceof Equal
-                    && predicate.literals().get(0) instanceof BinaryString) {
-                String equalValue = predicate.literals().get(0).toString();
-                if (tagManager.tagExists(equalValue)) {
-                    nameToSnapshot.put(equalValue, tagManager.tag(equalValue));
+                if (predicate instanceof CompoundPredicate) {
+                    CompoundPredicate compoundPredicate = (CompoundPredicate) predicate;
+                    // optimize for IN filter
+                    if ((compoundPredicate.function()) instanceof Or) {
+                        List<Predicate> children = compoundPredicate.children();
+                        for (Predicate leaf : children) {
+                            if (leaf instanceof LeafPredicate
+                                    && (((LeafPredicate) leaf).function() instanceof Equal)
```

Comment:
Why here not judge:
((LeafPredicate) predicate).literals().get(0) instanceof BinaryString
?